File: gems/decomplex/lib/decomplex/structural_topology.rb
# frozen_string_literal: true

require_relative "syntax"

module Decomplex
  # StructuralTopology is Decomplex's conservative static model of method
  # ownership and direct internal calls over Syntax structural facts. It
  # deliberately resolves only same-owner bare/self calls; dynamic dispatch
  # belongs to higher-recall detectors.
  class StructuralTopology
    Method = Struct.new(:id, :owner, :name, :file, :line, :span, :visibility, keyword_init: true)
    Edge = Struct.new(
      :caller, :callee, :caller_name, :callee_name, :file, :line, :span, :type, :kind, :confidence,
      keyword_init: true
    )

    def self.scan(files)
      documents = files.to_h do |file|
        [file, Syntax.parse(file, parser: "tree_sitter")]
      end

      methods = documents.flat_map do |file, document|
        MethodFacts.new(file, document).methods
      end
      edges = documents.flat_map do |file, document|
        EdgeFacts.new(file, document, methods).edges
      end
      edges.uniq! { |edge| [edge.caller, edge.callee, edge.type] }

      Graph.new(methods, edges)
    end

    class Graph
      attr_reader :methods, :edges

      def initialize(methods, edges)
        @methods = methods
        @edges = edges
        @method_by_id = methods.to_h { |method| [method.id, method] }
        @methods_by_owner = methods.group_by(&:owner)
        @edges_by_caller = edges.group_by(&:caller)
        @edges_by_callee = edges.group_by(&:callee)
        @edges_by_owner = edges.group_by { |edge| method(edge.caller)&.owner }
      end

      def method(id)
        @method_by_id[id]
      end

      def method_id(owner, name)
        "#{owner}##{name}"
      end

      def method_for(owner, name)
        method(method_id(owner, name))
      end

      def methods_for_owner(owner)
        Array(@methods_by_owner[owner])
      end

      def edges_for_owner(owner)
        Array(@edges_by_owner[owner])
      end

      def internal_calls(id)
        Array(@edges_by_caller[id])
      end

      def internal_callers(id)
        Array(@edges_by_callee[id])
      end

      def single_internal_caller?(id)
        internal_callers(id).map(&:caller).uniq.size == 1
      end

      def visibility(id)
        method(id)&.visibility
      end

      def owner(id)
        method(id)&.owner
      end

      def span(id)
        method(id)&.span
      end

      def call_sites(id)
        internal_calls(id).map do |edge|
          "#{edge.file}:#{edge.caller_name}:#{edge.line}"
        end
      end
    end

    class MethodFacts
      def initialize(file, document)
        @file = file
        @document = document
      end

      def methods
        @document.function_defs.map do |function|
          owner = owner_for_fact(function)
          Method.new(
            id: "#{owner}##{function.name}",
            owner: owner,
            name: function.name,
            file: @file,
            line: function.line,
            span: function.span,
            visibility: function.visibility || :public
          )
        end
      end

      private

      def owner_for_fact(fact)
        TopLevelOwner.new(@file, @document).owner_for(fact)
      end
    end

    class EdgeFacts
      def initialize(file, document, methods)
        @file = file
        @document = document
        @method_by_id = methods.to_h { |method| [method.id, method] }
        @owner_mapper = TopLevelOwner.new(file, document)
      end

      def edges
        @document.call_sites.filter_map do |call|
          edge_for_call(call)
        end
      end

      private

      def edge_for_call(call)
        return nil unless call.receiver.to_s == "self"

        owner = @owner_mapper.owner_for(call)
        caller = @method_by_id["#{owner}##{call.function}"]
        return nil unless caller

        callee_name = scoped_name(caller, call.message)
        callee = @method_by_id["#{owner}##{callee_name}"]
        return nil unless callee
        return nil if caller.id == callee.id

        Edge.new(
          caller: caller.id,
          callee: callee.id,
          caller_name: caller.name,
          callee_name: callee.name,
          file: @file,
          line: call.line,
          span: call.span,
          type: edge_type(call.control),
          kind: call_kind(call),
          confidence: :high
        )
      end

      def scoped_name(caller, message)
        caller.name.to_s.start_with?("self.") ? "self.#{message}" : message.to_s
      end

      def edge_type(control)
        %i[conditional iterates].include?(control) ? control : :always
      end

      def call_kind(call)
        source_text(call.span).lstrip.start_with?("self.") ? :direct_self : :bare_internal
      end

      def source_text(span)
        return "" unless span

        first_line, first_column, last_line, last_column = span
        if first_line == last_line
          return @document.lines[first_line - 1].to_s[first_column...last_column].to_s
        end

        parts = []
        parts << @document.lines[first_line - 1].to_s[first_column..].to_s
        parts.concat(@document.lines[first_line...(last_line - 1)] || [])
        parts << @document.lines[last_line - 1].to_s[0...last_column].to_s
        parts.join
      end
    end

    class TopLevelOwner
      def initialize(file, document)
        @file = file
        @document = document
      end

      def owner_for(fact)
        owner = fact.owner.to_s
        return owner unless owner == file_owner
        return owner if enclosed_by_matching_owner?(fact)

        top_level_owner
      end

      private

      def enclosed_by_matching_owner?(fact)
        @document.owner_defs.any? do |owner|
          owner.name.to_s == fact.owner.to_s && encloses?(owner.span, fact.span)
        end
      end

      def encloses?(outer, inner)
        return false unless outer && inner

        starts_before = outer[0] < inner[0] || (outer[0] == inner[0] && outer[1] <= inner[1])
        ends_after = outer[2] > inner[2] || (outer[2] == inner[2] && outer[3] >= inner[3])
        starts_before && ends_after
      end

      def file_owner
        File.basename(@file.to_s, File.extname(@file.to_s))
      end

      def top_level_owner
        "(top-level:#{@file})"
      end
    end
  end
end

File: gems/decomplex/rust/src/decomplex/detectors/structural_topology.rs
use crate::decomplex::syntax::{self, CallSite, Document, FunctionDef, Language, Span};
use anyhow::Result;
use serde::Serialize;
use std::collections::{BTreeMap, BTreeSet};
use std::path::{Path, PathBuf};

#[derive(Clone, Debug, Serialize)]
pub struct StructuralTopologyReport {
    pub methods: Vec<Method>,
    pub edges: Vec<Edge>,
}

#[derive(Clone, Debug, Serialize)]
pub struct Method {
    pub id: String,
    pub owner: String,
    pub name: String,
    pub file: String,
    pub line: usize,
    pub span: Span,
    pub visibility: String,
}

#[derive(Clone, Debug, Serialize)]
pub struct Edge {
    pub caller: String,
    pub callee: String,
    pub caller_name: String,
    pub callee_name: String,
    pub file: String,
    pub line: usize,
    pub span: Span,
    pub r#type: String,
    pub kind: String,
    pub confidence: String,
}

pub fn scan_files(files: &[PathBuf], language: Language) -> Result<StructuralTopologyReport> {
    let documents = syntax::parse_files(files, language)?;
    Ok(scan_documents(&documents))
}

pub fn scan_documents(documents: &[Document]) -> StructuralTopologyReport {
    let mut methods = Vec::new();
    for document in documents {
        methods.extend(methods_for_document(document));
    }
    methods.sort_by(|a, b| {
        a.file
            .cmp(&b.file)
            .then_with(|| a.line.cmp(&b.line))
            .then_with(|| a.owner.cmp(&b.owner))
            .then_with(|| a.name.cmp(&b.name))
    });

    let method_by_id = methods
        .iter()
        .map(|method| (method.id.clone(), method.clone()))
        .collect::<BTreeMap<_, _>>();

    let mut edges = Vec::new();
    for document in documents {
        edges.extend(edges_for_document(document, &method_by_id));
    }
    edges.sort_by(|a, b| {
        a.file
            .cmp(&b.file)
            .then_with(|| a.line.cmp(&b.line))
            .then_with(|| a.span.cmp(&b.span))
            .then_with(|| a.caller.cmp(&b.caller))
            .then_with(|| a.callee.cmp(&b.callee))
    });
    let mut seen = BTreeSet::new();
    edges.retain(|edge| {
        seen.insert((
            edge.caller.clone(),
            edge.callee.clone(),
            edge.r#type.clone(),
        ))
    });

    StructuralTopologyReport { methods, edges }
}

pub struct Graph {
    pub methods: Vec<Method>,
    pub edges: Vec<Edge>,
    method_by_id: BTreeMap<String, Method>,
    edges_by_caller: BTreeMap<String, Vec<Edge>>,
    edges_by_callee: BTreeMap<String, Vec<Edge>>,
}

impl Graph {
    pub fn new(methods: Vec<Method>, edges: Vec<Edge>) -> Self {
        let mut method_by_id = BTreeMap::new();
        for m in &methods {
            method_by_id.insert(m.id.clone(), m.clone());
        }

        let mut edges_by_caller = BTreeMap::new();
        let mut edges_by_callee = BTreeMap::new();
        for e in &edges {
            edges_by_caller
                .entry(e.caller.clone())
                .or_insert_with(Vec::new)
                .push(e.clone());
            edges_by_callee
                .entry(e.callee.clone())
                .or_insert_with(Vec::new)
                .push(e.clone());
        }

        Self {
            methods,
            edges,
            method_by_id,
            edges_by_caller,
            edges_by_callee,
        }
    }

    pub fn method(&self, id: &str) -> Option<&Method> {
        self.method_by_id.get(id)
    }

    pub fn internal_calls(&self, id: &str) -> Vec<Edge> {
        self.edges_by_caller.get(id).cloned().unwrap_or_default()
    }

    pub fn internal_callers(&self, id: &str) -> Vec<Edge> {
        self.edges_by_callee.get(id).cloned().unwrap_or_default()
    }

    pub fn single_internal_caller(&self, id: &str) -> bool {
        let callers = self.internal_callers(id);
        let mut unique = BTreeMap::new();
        for c in callers {
            unique.insert(c.caller, true);
        }
        unique.len() == 1
    }

    pub fn visibility(&self, id: &str) -> Option<&str> {
        self.method(id).map(|m| m.visibility.as_str())
    }
}

fn methods_for_document(document: &Document) -> Vec<Method> {
    document
        .function_defs
        .iter()
        .map(|function| method_for_function(document, function))
        .collect()
}

fn method_for_function(document: &Document, function: &FunctionDef) -> Method {
    let owner = top_level_owner_for(document, &function.owner, function.span);
    Method {
        id: format!("{}#{}", owner, function.name),
        owner,
        name: function.name.clone(),
        file: function.file.clone(),
        line: function.line,
        span: function.span,
        visibility: function
            .visibility
            .clone()
            .unwrap_or_else(|| "public".to_string()),
    }
}

fn edges_for_document(document: &Document, method_by_id: &BTreeMap<String, Method>) -> Vec<Edge> {
    document
        .call_sites
        .iter()
        .filter_map(|call| edge_for_call(document, method_by_id, call))
        .collect()
}

fn edge_for_call(
    document: &Document,
    method_by_id: &BTreeMap<String, Method>,
    call: &CallSite,
) -> Option<Edge> {
    if call.receiver != "self" {
        return None;
    }

    let owner = top_level_owner_for(document, &call.owner, call.span);
    let caller = method_by_id.get(&format!("{}#{}", owner, call.function))?;
    let dialect = crate::decomplex::dialect::dialect_for_document(document);
    let callee_name = dialect.scoped_name(&caller.name, &call.message);
    let callee = method_by_id.get(&format!("{}#{}", owner, callee_name))?;
    if caller.id == callee.id {
        return None;
    }

    Some(Edge {
        caller: caller.id.clone(),
        callee: callee.id.clone(),
        caller_name: caller.name.clone(),
        callee_name: callee.name.clone(),
        file: call.file.clone(),
        line: call.line,
        span: call.span,
        r#type: edge_type(call.control.as_deref()),
        kind: call_kind(call),
        confidence: "high".to_string(),
    })
}

fn edge_type(control: Option<&str>) -> String {
    match control {
        Some("conditional" | "iterates") => control.unwrap().to_string(),
        _ => "always".to_string(),
    }
}

fn call_kind(call: &CallSite) -> String {
    if call.receiver == "self" {
        "internal_self".to_string()
    } else {
        "internal".to_string()
    }
}

fn top_level_owner_for(document: &Document, owner: &str, span: Span) -> String {
    if owner != file_owner(&document.file) || enclosed_by_matching_owner(document, owner, span) {
        owner.to_string()
    } else {
        format!("(top-level:{})", document.file)
    }
}

fn file_owner(file: &str) -> String {
    Path::new(file)
        .file_stem()
        .and_then(|stem| stem.to_str())
        .filter(|stem| !stem.is_empty())
        .unwrap_or("(file)")
        .to_string()
}

fn enclosed_by_matching_owner(document: &Document, owner: &str, span: Span) -> bool {
    document
        .owner_defs
        .iter()
        .any(|owner_def| owner_def.name == owner && span_encloses(owner_def.span, span))
}

fn span_encloses(outer: Span, inner: Span) -> bool {
    let starts_before = outer[0] < inner[0] || (outer[0] == inner[0] && outer[1] <= inner[1]);
    let ends_after = outer[2] > inner[2] || (outer[2] == inner[2] && outer[3] >= inner[3]);
    starts_before && ends_after
}
